### PR TITLE
Added documentation anchors for tutorial and cookbook

### PR DESCRIPTION
**Pull request by mcfadden8**
This addresses issue #429 

All references by line number in the documentation have been updated to use anchors instead.  This will protect us from future code style changes making the line numbers get out of sync with what is being displayed.

Updated documentation for this branch may be reviewed here: https://umpire.readthedocs.io/en/bugfix-um-762-source-examples-in-tutorial-documentation-need-fixup-post-style-updates/

### CODE REVIEW

**Comment by mcfadden8** [comment]
LGTM

**Comment by mcfadden8** [comment]
LGTM

**Comment by mcfadden8** [comment]
LGTM

**Comment by mcfadden8** [comment]
LGTM

**Comment by mcfadden8** [comment]
LGTM

File: docs/sphinx/tutorial/allocators.rst [hunk 1]
```diff
@@ -13,14 +13,19 @@ All :class:`umpire::Allocator` s are created and managed by Umpire's
...
-                    :lines: 14-16
+   :start-after: _umpire_tut_get_allocator_start
+   :end-before: _umpire_tut_get_allocator_end
+   :language: C++
 
 Once you have an :class:`umpire::Allocator` you can use it to allocate and deallocate memory:
 
 .. literalinclude:: ../../../examples/tutorial/tut_allocator.cpp
-                    :lines: 18-24
+   :start-after: _umpire_tut_de_allocate_start
```
Comment (kab163) [inline docs/sphinx/tutorial/allocators.rst:23]: This section of the code shows getting a size, allocating memory, printing out what you have, and then deallocating. What do you think about shaving it down to just allocate and deallocating? (Or perhaps just what the original code snippet was - the allocation, print statement, and deallocating). Do you see what I mean?

**Comment by kab163** [comment]
Is the purpose of comments like "// _umpire_tut_get_allocator_start", etc. just to show a reader/user which lines were called out earlier on the page? Like with the Allocators page in the Tutorial section, there are two code snippets called out before the whole example code is shown. So I assume the purpose of the comments in the complete example is just to show where the code pieces came from?

File: docs/sphinx/tutorial/allocators.rst (hunk 1 above)
Comment (kab163) [reply docs/sphinx/tutorial/allocators.rst:23]: Similar question for the Dynamic Pools Tutorial page

File: docs/sphinx/cookbook/move_to_managed.rst [hunk 2]
```diff
@@ -9,7 +9,9 @@ should be moved to unified memory in order to make it accessible by the GPU.
 You can do this with the :func:`umpire::ResourceManager::move` operation:
 
 .. literalinclude:: ../../../examples/cookbook/recipe_move_to_managed.cpp
-                    :lines: 26
+   :start-after: _umpire_tut_move_host_to_managed_start
```
Comment (kab163) [inline docs/sphinx/cookbook/move_to_managed.rst:12]: This is another spot where I think too much code is included in the snippet. What do you think about only including the code below the "// Move data to unified memory" comment?

File: examples/tutorial/c/tut_allocator.c [hunk 3]
```diff
@@ -11,17 +11,21 @@
 #define SIZE 1024
 
 int main() {
+  /* _umpire_tut_c_get_allocator_start */
```
Comment (kab163) [inline examples/tutorial/c/tut_allocator.c:14]: Why the change in comment formatting for the C examples?

Comment (kab163) [reply examples/tutorial/c/tut_allocator.c:14]: Lol I guess it doesn't matter that much anyways, but hey

**Comment by mcfadden8** [comment]
> Is the purpose of comments like "// _umpire_tut_get_allocator_start", etc. just to show a reader/user which lines were called out earlier on the page? Like with the Allocators page in the Tutorial section, there are two code snippets called out before the whole example code is shown. So I assume the purpose of the comments in the complete example is just to show where the code pieces came from?

The comments have to be in the code so that sphynx will know which code snippet to show.  The comments are actually tags that sphynx uses.  Does this make sense?  Previously, we used line numbers which turned out to be problematic when code got shifted around.

I suppose we could rename the tags to something that would indicate they are required by the documentation tools?

File: examples/tutorial/c/tut_allocator.c (hunk 3 above)
Comment (mcfadden8) [reply examples/tutorial/c/tut_allocator.c:14]: I was just trying to make it obvious to the viewer that they were seeing C versus C++ versus FORTRAN versus JSON.  We can change it back if it really causes pain.

**Comment by kab163** [comment]
> > Is the purpose of comments like "// _umpire_tut_get_allocator_start", etc. just to show a reader/user which lines were called out earlier on the page? Like with the Allocators page in the Tutorial section, there are two code snippets called out before the whole example code is shown. So I assume the purpose of the comments in the complete example is just to show where the code pieces came from?
> 
> The comments have to be in the code so that sphynx will know which code snippet to show. The comments are actually tags that sphynx uses. Does this make sense? Previously, we used line numbers which turned out to be problematic when code got shifted around.
> 
> I suppose we could rename the tags to something that would indicate they are required by the documentation tools?

It's up to you guys, I just didn't realize it was a Sphinx thing. I see what's going on now - at first, it looks like simple comments, but yea that makes more sense.

File: examples/tutorial/c/tut_allocator.c (hunk 3 above)
Comment (kab163) [reply examples/tutorial/c/tut_allocator.c:14]: Ah, I see. Again, it's up to you guys, I was just kinda curious. I think it's fine either way!

File: docs/sphinx/tutorial/allocators.rst (hunk 1 above)
Comment (mcfadden8) [reply docs/sphinx/tutorial/allocators.rst:23]: Thanks @kab163, can you enter this as a suggestion (lille +/- box) at the top?

**Comment by mcfadden8** [comment]
> > > Is the purpose of comments like "// _umpire_tut_get_allocator_start", etc. just to show a reader/user which lines were called out earlier on the page? Like with the Allocators page in the Tutorial section, there are two code snippets called out before the whole example code is shown. So I assume the purpose of the comments in the complete example is just to show where the code pieces came from?
> > 
> > 
> > The comments have to be in the code so that sphynx will know which code snippet to show. The comments are actually tags that sphynx uses. Does this make sense? Previously, we used line numbers which turned out to be problematic when code got shifted around.
> > I suppose we could rename the tags to something that would indicate they are required by the documentation tools?
> 
> It's up to you guys, I just didn't realize it was a Sphinx thing. I see what's going on now - at first, it looks like simple comments, but yea that makes more sense.

It probably would make more sense to rename the tags somehow to tell users they should not be removed and should be moved with care.

File: examples/tutorial/tut_allocator.cpp [hunk 4]
```diff
@@ -9,11 +9,14 @@
...
-  constexpr std::size_t SIZE = 1024;
-
+  // _umpire_tut_get_allocator_start
   auto& rm = umpire::ResourceManager::getInstance();
 
   umpire::Allocator allocator = rm.getAllocator("HOST");
+  // _umpire_tut_get_allocator_end
+
+  // _umpire_tut_de_allocate_start
+  constexpr std::size_t SIZE = 1024;
```
Comment (kab163) [inline examples/tutorial/tut_allocator.cpp:20]:
```suggestion
// _umpire_tut_de_allocate_start
```

File: examples/tutorial/tut_allocator.cpp [hunk 5]
```diff
@@ -9,11 +9,14 @@
...
 {
-  constexpr std::size_t SIZE = 1024;
-
+  // _umpire_tut_get_allocator_start
   auto& rm = umpire::ResourceManager::getInstance();
 
   umpire::Allocator allocator = rm.getAllocator("HOST");
+  // _umpire_tut_get_allocator_end
+
+  // _umpire_tut_de_allocate_start
```
Comment (kab163) [inline examples/tutorial/tut_allocator.cpp:18]:
```suggestion

```

**Comment by kab163** [comment]
> > > > Is the purpose of comments like "// _umpire_tut_get_allocator_start", etc. just to show a reader/user which lines were called out earlier on the page? Like with the Allocators page in the Tutorial section, there are two code snippets called out before the whole example code is shown. So I assume the purpose of the comments in the complete example is just to show where the code pieces came from?
> > > 
> > > 
> > > The comments have to be in the code so that sphynx will know which code snippet to show. The comments are actually tags that sphynx uses. Does this make sense? Previously, we used line numbers which turned out to be problematic when code got shifted around.
> > > I suppose we could rename the tags to something that would indicate they are required by the documentation tools?
> > 
> > 
> > It's up to you guys, I just didn't realize it was a Sphinx thing. I see what's going on now - at first, it looks like simple comments, but yea that makes more sense.
> 
> It probably would make more sense to rename the tags somehow to tell users they should not be removed and should be moved with care.

Very true. I think that's a good idea

File: examples/tutorial/tut_allocator.cpp (hunk 5 above)
Comment (mcfadden8) [reply examples/tutorial/tut_allocator.cpp:18]: I think I see what you are asking for.  Let me see if I can try something.

Comment (kab163) [reply examples/tutorial/tut_allocator.cpp:18]: I'm sorry, I haven't used this suggestion feature thingy before. Basically, I think moving the "// _umpire_tut_de_allocate_start" statement from line 18 to line 20 or so will make it only include the allocate + print statement + deallocate. (Kinda narrows the focus to only the allocate/deallocate which is what the description for the code snippet is talking about anyway).

**Comment by mcfadden8** [comment]
LGTM

**Comment by mcfadden8** [comment]
> > > > > Is the purpose of comments like "// _umpire_tut_get_allocator_start", etc. just to show a reader/user which lines were called out earlier on the page? Like with the Allocators page in the Tutorial section, there are two code snippets called out before the whole example code is shown. So I assume the purpose of the comments in the complete example is just to show where the code pieces came from?
> > > > 
> > > > 
> > > > The comments have to be in the code so that sphynx will know which code snippet to show. The comments are actually tags that sphynx uses. Does this make sense? Previously, we used line numbers which turned out to be problematic when code got shifted around.
> > > > I suppose we could rename the tags to something that would indicate they are required by the documentation tools?
> > > 
> > > 
> > > It's up to you guys, I just didn't realize it was a Sphinx thing. I see what's going on now - at first, it looks like simple comments, but yea that makes more sense.
> > 
> > 
> > It probably would make more sense to rename the tags somehow to tell users they should not be removed and should be moved with care.
> 
> Very true. I think that's a good idea

I added a "_sphinx_doc_" prefix to all of the tags.

File: examples/tutorial/tut_allocator.cpp (hunk 5 above)
Comment (mcfadden8) [reply examples/tutorial/tut_allocator.cpp]: I've done this.  Thanks for your comments.  It will take about ten minutes for the document to be updated.  I appreciate your review and have narrowed the focus where you suggested.

**Review by davidbeckingsale — APPROVED** [review]
This is awesome, thanks @mcfadden8 !

**Comment by mcfadden8** [comment]
LGTM

**Comment by mcfadden8** [comment]
Thanks @davidbeckingsale, do you know why I can't merge?  I have no idea why LLNL.Umpire is stuck?
